protocols/mp-ecdsa/src/protocols/managers/keygen.rs

```diff
@@ -0,0 +1,554 @@
+            } else {
+                KeygenRound::Next
+            };
+
+            return self.maybe_start_keygen_for_stage(stage, header, anticipated_execution)
+                .await;
+        }
+
+        // It's possible genesis failed and we need to retry
+        if session_id == GENESIS_AUTHORITY_SET_ID
+            && matches!(state, KeygenState::Failed { session_id: 0 })
+            && self.client.dkg_pub_key_is_unset(header).await
+        {
+            self
+                .logger
+                .warn("We will trigger another genesis keygen because the previous one failed");
+            return self.maybe_start_keygen_for_stage(
+                KeygenRound::Genesis,
+                header,
+                anticipated_execution,
+            )
+            .await;
+        }
+
+        None
+    }
+
+    /// Check to see if we need to run a genesis keygen (session = 0), or, a keygen for session 1
+    async fn genesis_checks(
+        &self,
+        state: KeygenState,
+        header: &B::Header,
+        anticipated_execution: &AnticipatedKeygenExecutionStatus,
+    ) -> Option<Job> {
+        if state == KeygenState::Uninitialized {
+            // If we are at genesis, and there is no active keygen, create and immediately
+            // start() one
+            return self
+                .maybe_start_keygen_for_stage(
+                    KeygenRound::Genesis,
+                    header,
+                    anticipated_execution,
+                )
+                .await;
+        }
+
+        if state == KeygenState::RunningGenesisKeygen {
+            // If we are at genesis, and a genesis keygen is running, do nothing
+            return None
+        }
+
+        if state == KeygenState::RunningKeygen {
+            // If we are at genesis, and there is a next keygen running, do nothing
+            return None
+        }
+
+        if matches!(
+            state,
+            KeygenState::KeygenCompleted {
+                session_completed: 0
+            }
+        ) {
+            // If we are at genesis, and we have completed keygen, we may need to begin a keygen
+            // for session 1
+            return self
+                .maybe_start_keygen_for_stage(
+                    KeygenRound::Next,
+                    header,
+                    anticipated_execution,
+                )
+                .await;
+        }
+
+        if matches!(state, KeygenState::Failed { session_id: 1 }) {
+            // If we are at genesis, and we have failed keygen for session 1, we may need to begin a
+            // keygen for session 1
+            return self
+                .maybe_start_keygen_for_stage(
+                    KeygenRound::Next,
+                    header,
+                    anticipated_execution,
+                )
+                .await;
+        }
+
+        None
+    }
+
+    /// Check to see if we need to run a keygen for session 2, 3, 4, .., etc.
```

Comment:
This isn't a thing anymore. If there's confusion here you need to clarify how we can assist.